### DISCUSSION TITLE
VipsOperation: class "heifsave_buffer" not found

### DISCUSSION THREAD

**Discussion by At4m4n** [libvips/libvips#2340] [Q&A] [answered]
## Bug report

**Describe the bug**
Using AWS Lambda layer to configure the custom nodejs12.x runtime with sharp and libvips. I use Makefile for layer to install all needed dependencies including libvips, libheif, libde265, x265 encoder and other dev packages. Layer builds fine, everything seems to be installed, but when call the sharp js code, receive "VipsOperation: class \"heifsave_buffer\" not found" error message.

**To Reproduce**
(js code and template are omitted)
1. Create lambda project folder (e.g. via `sam init`)
2. Reference local layer for your function in template
3. Setup layer with following Makefile:
```
# Library Versions
LIBDE265_VERSION=1.0.8
LIBHEIF_VERSION=1.12.0
VIPS_VERSION=8.11.2
SHARP_VERSION=0.28.3

PREFIX_PATH=/opt

export PKG_CONFIG_PATH=$(PREFIX_PATH)/lib/pkgconfig

build-SharpHEICLayer: libvips
	mkdir -p "$(ARTIFACTS_DIR)/nodejs"
	mkdir -p "$(ARTIFACTS_DIR)/lib"

	LD_LIBRARY_PATH=$(PREFIX_PATH)/lib npm --prefix "$(ARTIFACTS_DIR)/nodejs/" install sharp@$(SHARP_VERSION)
	ldd $(ARTIFACTS_DIR)/nodejs/node_modules/sharp/build/Release/sharp.node | sed -nE "s/^[^\/\n]*(\/[^ ]+)(.*)/\1/p" | xargs cp -t $(ARTIFACTS_DIR)/lib

x265:
	curl -L http://li.nux.ro/download/nux/dextop/el7/x86_64/nux-dextop-release-0-5.el7.nux.noarch.rpm --output nux-dextop
	yum -y install https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm
	rpm -Uvh nux-dextop
	yum install -y x265

libde265: x265
	curl -L https://github.com/strukturag/libde265/releases/download/v$(LIBDE265_VERSION)/libde265-$(LIBDE265_VERSION).tar.gz | tar zx
	cd libde265-$(LIBDE265_VERSION) && ./autogen.sh && ./configure --disable-dec265 --disable-sherlock265 --prefix=$(PREFIX_PATH) && make V=1 && make install

libheif: libde265
	curl -L https://github.com/strukturag/libheif/releases/download/v$(LIBHEIF_VERSION)/libheif-$(LIBHEIF_VERSION).tar.gz | tar zx
	cd libheif-$(LIBHEIF_VERSION) && ./autogen.sh && ./configure --prefix=$(PREFIX_PATH) && make V=1 && make install

libvips: installLib libheif
	curl -L https://github.com/libvips/libvips/releases/download/v$(VIPS_VERSION)/vips-$(VIPS_VERSION).tar.gz | tar zx
	cd vips-$(VIPS_VERSION) && ./configure --prefix=$(PREFIX_PATH) && make V=1 && make install

installLib:
	yum install -y \
		glib2-devel \
		expat-devel \
		lcms2-devel \
		libjpeg-devel \
		libjpeg-turbo-devel \
		libpng-devel \
		giflib-devel \
		libexif-devel \
		librsvg2-devel \
		libtiff-devel
```
4. Invoke sharp code on any .heic image (e.g. via `sam local invoke`)

**Expected behavior**
Sharp processes image without errors.

**Actual behavior**
Error with message VipsOperation: class \"heifsave_buffer\" not found

**Environment**
amazon/aws-sam-cli-build-image-nodejs12.x:latest
 - OS: Amazon Linux 2
 - Vips: 8.11.2

**Additional context**
Also, opened issue on [sharp side](https://github.com/lovell/sharp/issues/2783).

**Comment by jcupitt** [comment]
Hi @At4m4n,

Could you try running your test with the env var `VIPS_INFO=1`? It should produce some output describing how it's searching for run-time modules. For example, I see:

```
$ VIPS_INFO=1 vips
VIPS-INFO: 13:54:53.483: found /home/john/vips/lib/vips-modules-8.11
VIPS-INFO: 13:54:53.483: using configure-time prefix
VIPS-INFO: 13:54:53.483: VIPS_PREFIX = /home/john/vips
VIPS-INFO: 13:54:53.483: VIPS_LIBDIR = /home/john/vips/lib
VIPS-INFO: 13:54:53.483: prefix = /home/john/vips
VIPS-INFO: 13:54:53.483: libdir = /home/john/vips/lib
VIPS-INFO: 13:54:53.483: searching "/home/john/vips/lib/vips-modules-8.11"
VIPS-INFO: 13:54:53.483: loading "/home/john/vips/lib/vips-modules-8.11/vips-openslide.so"
VIPS-INFO: 13:54:53.484: loading "/home/john/vips/lib/vips-modules-8.11/vips-heif.so"
VIPS-INFO: 13:54:53.489: loading "/home/john/vips/lib/vips-modules-8.11/vips-poppler.so"
VIPS-INFO: 13:54:53.492: loading "/home/john/vips/lib/vips-modules-8.11/vips-magick.so"
VIPS-INFO: 13:54:53.493: searching "/home/john/vips/lib/vips-plugins-8.11"
Usage:
...
```

ie. it's loaded the heif module from `/home/john/vips/lib/vips-modules-8.11`.

**Comment by At4m4n** [comment]
Hi @jcupitt . This is what I get during function execution with that flag set to 1
```
VIPS-INFO: 13:20:02.389: g_getenv( "PATH" ) == "/var/lang/bin:/usr/local/bin:/usr/bin/:/bin:/opt/bin"
VIPS-INFO: 13:20:02.390: looking in "/var/lang/bin" for "sharp"
VIPS-INFO: 13:20:02.390: looking in "/usr/local/bin" for "sharp"
VIPS-INFO: 13:20:02.390: looking in "/usr/bin/" for "sharp"
VIPS-INFO: 13:20:02.390: looking in "/bin" for "sharp"
VIPS-INFO: 13:20:02.390: looking in "/opt/bin" for "sharp"
VIPS-INFO: 13:20:02.390: trying for dir = "/var/task/sharp", name = "sharp"
VIPS-INFO: 13:20:02.390: canonicalised path = "/var/task"
VIPS-INFO: 13:20:02.390: VIPS_PREFIX = /opt
VIPS-INFO: 13:20:02.390: VIPS_LIBDIR = /opt/lib
VIPS-INFO: 13:20:02.390: prefix = /opt
VIPS-INFO: 13:20:02.390: libdir = /opt/lib
VIPS-INFO: 13:20:02.392: searching "/opt/lib/vips-plugins-8.11"
<some js exectuion logs about start of image processing>
VIPS-INFO: 13:20:07.254: shrinkv by 13
VIPS-INFO: 13:20:07.255: shrinkh by 13
VIPS-INFO: 13:20:07.255: residual reducev by 0.483631
VIPS-INFO: 13:20:07.255: reducev: 13 point mask
VIPS-INFO: 13:20:07.256: residual reduceh by 0.483631
VIPS-INFO: 13:20:07.256: reduceh: 13 point mask
<function result with error message>
```

**Comment by jcupitt** [comment]
It looks like your libvips has built without module support, so heifsave should be part of the base library. It's obviously not there, so I think libvips configure must have failed to find libheif-devel. 

What do you see for the table at the end of libvips configure? You ought to have something like:

```
image pyramid save:                     yes
 (requires libgsf-1 1.14.26 or later)
HEIC/AVIF load/save with libheif:       yes (dynamic module: yes)
WebP load/save with libwebp:            yes
```

If you see `no` for libheif you've probably not set `PKG_CONFIG_PATH`. Try adding:

```
export PKG_CONFIG_PATH=$(PREFIX)/lib/pkgconfig:$PKG_CONFIG_PATH
```

before you run libvips configure.

**Comment by jcupitt** [comment]
Oh wait, I see you have that. I suppose something else is stopping libvips finding libheif.

**Comment by jcupitt** [comment]
(removed bug and turned to discussion, since this seems to be an install issue)

**Comment by kleisauke** [comment]
fwiw, you could also you try running `vips --vips-config` to also check whether libvips was successfully built against `libheif`.

Note that libvips disables the HEIC/AVIF encoder/decoder once these commands returns both `no`:
```bash
# Encoder
$ pkg-config libheif --variable builtin_h265_encoder
$ pkg-config libheif --variable builtin_avif_encoder

# Decoder
$ pkg-config libheif --variable builtin_h265_decoder
$ pkg-config libheif --variable builtin_avif_decoder
```

**Reply by At4m4n** [reply]
`vips --vips-config` output
```
enable debug: no
enable deprecated library components: yes
enable modules: no
use fftw3 for FFT: no
accelerate loops with orc: no
ICC profile support with lcms: yes (lcms2)
zlib: yes
text rendering with pangocairo: no
font file support with fontconfig:
RAD load/save: yes
Analyze7 load/save: yes
PPM load/save: yes
GIF load:  yes
EXIF metadata support with libexif: yes
JPEG load/save with libjpeg: yes (pkg-config)
JXL load/save with libjxl: no (dynamic module: no)
JPEG2000 load/save with libopenjp2: no
PNG load with libspng: no
PNG load/save with libpng: yes (pkg-config libpng >= 1.2.9)
PNG quantisation to 8 bit: no
TIFF load/save with libtiff: yes (pkg-config libtiff-4)
image pyramid save: no
HEIC/AVIF load/save with libheif: no (dynamic module: no)
WebP load/save with libwebp: no
PDF load with PDFium:  no
PDF load with poppler-glib: no (dynamic module: no)
SVG load with librsvg-2.0: yes
EXR load with OpenEXR: no
OpenSlide load: no (dynamic module: no)
Matlab load with matio: no
NIfTI load/save with niftiio: no
FITS load/save with cfitsio: no
Magick package: none (dynamic module: no)
Magick API version: none
load with libMagickCore: no
save with libMagickCore: no
```
Looks like vips can't see modules installed with make and make install, unlike those installed with yum (lcms2, for example)

**Reply by At4m4n** [reply]
Btw, libheif sees both encoder and decoder, so the problem is to make vips see the libheif.
```
$ pkg-config libheif --variable builtin_h265_encoder
yes
$ pkg-config libheif --variable builtin_h265_decoder
yes
```

**Reply by At4m4n** [reply]
managed to install libheif-devel, now in vips config I see this
`HEIC/AVIF load/save with libheif: yes (dynamic module: no)`
but still the same error

**Reply by jcupitt** [reply]
Then I suppose you are missing the encoder. `vips -l | grep heif` will list all the heif-related operations that are enabled.

**Reply by At4m4n** [reply]
Yes, you're right. It shows only `heifload_buffer`, but not `heifsave_buffer`. 
```
vips -l | grep heif
        VipsForeignLoadHeif (heifload_base), load a HEIF image, priority=0
          VipsForeignLoadHeifFile (heifload), load a HEIF image (.heic, .heif, .avif), priority=0, is_a, get_flags, header, load
          VipsForeignLoadHeifBuffer (heifload_buffer), load a HEIF image, priority=0, is_a_buffer, get_flags, header, load
          VipsForeignLoadHeifSource (heifload_source), load a HEIF image, priority=0, is_a_source, get_flags, header, load
```
Does it mean it sees the decoder, but not encoder?

**Comment by lovell** [comment]
@At4m4n Please don't cross-post. https://github.com/lovell/sharp/issues/2783

**Reply by jcupitt** [reply]
Sorry Lovell, I hadn't seen you were already looking at this, I'll do something else.

Yes, @At4m4n, please don't cross-post, it makes unnecessary extra work.

**Reply by At4m4n** [reply]
Sorry for that, wasn't sure whhich project is more realted to the problem I have. Now it looks like libvips installation is wrong and it can't see some dependencies. Let's continue in this thread, I'll close sharp's issue

**Comment by jcupitt** [comment]
As Kleis says, libvips configure uses this to detect libheif:

```
$ pkg-config libheif --print-variables
builtin_avif_decoder
builtin_avif_encoder
builtin_h265_decoder
builtin_h265_encoder
exec_prefix
includedir
libdir
pcfiledir
prefix
```

Presumably pkg-config can't see libheif, or can't see the correct version. You'll need to investigate.

**Comment by kleisauke** [comment, marked as answer]
Is this for Amazon Linux 2? If so, you could also use Remi's RPM repository which provides an up-to-date version of libvips including optional HEIC support.
```bash
# Install and enable the EPEL RPM package on Amazon Linux 2
yum install https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm

# Install the RPM Fusion configuration package
yum install https://download1.rpmfusion.org/free/el/rpmfusion-free-release-7.noarch.rpm

# Install the Remi repository configuration package
yum install http://rpms.remirepo.net/enterprise/remi-release-7.rpm

# Install the yum-utils package (for the yum-config-manager command)
yum install yum-utils

# Enable Remi's RPM repository
yum-config-manager --enable remi

# Install libvips with HEIC support (+ development files and command-line tools)
yum install vips vips-heif vips-devel vips-tools
```

See also: https://github.com/libvips/libvips/issues/1184#issuecomment-447973135.

**Reply by At4m4n** [reply]
@kleisauke Yes it is. Looks promising, will check soon.

**Reply by kleisauke** [reply]
Unfortunately, RPM Fusion provides somewhat outdated versions for libheif, libde265 and x265 on EL7, see:
https://pkgs.rpmfusion.org/cgit/free/libheif.git/tree/libheif.spec?h=el7#n2
https://pkgs.rpmfusion.org/cgit/free/libde265.git/tree/libde265.spec?h=el7#n3
https://pkgs.rpmfusion.org/cgit/free/x265.git/tree/x265.spec?h=el7#n5

So it might be better to compile from source in this case (or by using a system based on RHEL 8).

**Reply by kleisauke** [reply]
Oh, I just noticed that it doesn't compile against libaom either on EL7, so it doesn't include AVIF support.

**Reply by At4m4n** [reply]
It's not that important for me now, actually. Versions look not too old. AVIF support is not a primary task. Just need to be able to process .heic for now.
Yum install you suggested worked, vips config shows `yes` for almost everything
But sharp isn't still willing to process, getting `Input buffer contains unsupported image format` , which feels like sharp doesn't use vips after I install it with yum

**Reply by kleisauke** [reply]
Did you see these information lines while installing sharp?:
```
sharp: Detected globally-installed libvips v8.11.2
sharp: Building from source via node-gyp
```

**Reply by At4m4n** [reply]
yes

**Reply by At4m4n** [reply]
oh, I must've been blind yesterday and missed vips-heif dependency in yum install (#1184 comment doesn't have it, probably copied from there). So the errors with finding vips  and heif classes gone as I understand, at least in the container I test it in.
Big thanks for that!
Unfortunately, still can't have an output, getting `heifsave: Unsupported compression` on img.toBuffer() after resizing the image. vips-heif brought both x265 and libde265 transitively as I see it in yum output, but looks like something still can't cope with the file

**Reply by kleisauke** [reply]
No problem. Did you try to set the `.heif({ compression: 'hevc' })` option in sharp? Otherwise it will save AVIF files by default (which won't work on EL7).

**Reply by kleisauke** [reply]
(I'll make the `vips-heif` dependency more explicit next time, issue https://github.com/remicollet/remirepo/issues/171 contains some more details on this)

**Reply by At4m4n** [reply]
That setting worked! 
@kleisauke , @jcupitt , @lovell Appreciate your help and hard work on your projects!
